Notifications rework

jeabakker opened the pull request:
ToDo
- [x] tests
- [x] muting notifications

jeabakker committed:
chore(core): notification lib functions return value

jeabakker committed:
feat(notifications): more detailed notification subscriptions

jeabakker committed:
chore(core): constructor params

jeabakker committed:
chore(core): function param name

jeabakker committed:
chore(notifications): check registered notification methods

jdalsem commented on engine/classes/Elgg/Traits/Entity/Subscriptions.php:
```diff
@@ -0,0 +1,198 @@
+			$dbt = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2);
+			$caller = $dbt[1]['function'] ?? 'unknown';
+			
+			throw new InvalidParameterException(elgg_echo('Entity:Subscriptions:InvalidMethodsException', [$caller]));
```
change to invalidargumentexception (in whole file)

jeabakker replied:
done

jeabakker committed:
feat(subscriptions): added muting possibility to entity
subscriptions

jeabakker committed:
feat(site_notifications): site notifications are enabled for new users

jeabakker committed:
feat(users): uniform storage of notification settings

It's not longer needed to have your own storage method for different
notification preferences of a user

fixes #10185

jeabakker committed:
feat(core): users are automaticly subscribed to content they comment on

jeabakker committed:
feat(actions): added generic actions for (un)subscribe and (un)mute

jeabakker committed:
chore(core): fix typo in function name

jdalsem commented on engine/upgrades.php:
```diff
@@ -10,4 +10,5 @@
 	\Elgg\Upgrades\AlterDatabaseToMultiByteCharset::class,
 	\Elgg\Upgrades\DeleteDiagnosticsPlugin::class,
 	\Elgg\Upgrades\NotificationsPrefix::class,
+	\Elgg\Upgrades\ChangeUserNotificationSettingsNamespace::class,
```
should be placed alphabetically i believe

jdalsem commented on languages/en.php:
```diff
@@ -1657,7 +1666,19 @@
+	'entity:unsubscribe:success' => "You've successfully unsubscribed from %s",
+	'entity:unsubscribe:fail' => "An error occured while unsubscribing from %s",
+	
+	'entity:mute:success' => "You've successfully muted notifications about %s",
```
notifications about should be notifications of (multiple lines)

jdalsem commented on mod/friends/classes/Elgg/Friends/Relationships.php:
```diff
@@ -17,7 +17,7 @@ class Relationships {
 	 * @return void
 	 * @since 3.2
 	 */
-	public static function createFriendRelationship(\Elgg\Event $event) {
```
rename should be documented

jeabakker replied:
was documented

jdalsem commented on mod/groups/classes/Elgg/Groups/Relationships.php:
```diff
@@ -0,0 +1,68 @@
+	 *
+	 * @return void
+	 */
+	public static function removeGroupNotificationSubscription(\Elgg\Event $event): void {
```
function name should be plural subscriptions

jdalsem commented on docs/appendix/upgrade-notes/3.x-to-4.0.rst:
```diff
@@ -99,6 +99,19 @@ Multiple Recipients
+To display the setting extend the view ``notifications/settings/records`` (plural) with a view which uses ``notifications/settings/record`` (singular).
+
+When requesting notification settings other than the default setting, if the user hasn't saved a setting yet it'll fall back to the default notification settings.
+
```
No mention about removal of friends collections notification management

jdalsem commented on mod/notifications/elgg-plugin.php:
```diff
@@ -13,13 +15,19 @@
 		'settings:notification:personal' => [
 			'path' => '/notifications/personal/{username?}',
 			'resource' => 'notifications/index',
+			'middelware' => [
```
middleware

jdalsem commented on mod/notifications/elgg-plugin.php:
```diff
@@ -13,13 +15,19 @@
 			'required_plugins' => [
 				'groups',
 			],
+			'middelware' => [
```
middleware

jdalsem commented on mod/notifications/languages/en.php:
```diff
@@ -7,18 +7,16 @@
+	'notifications:subscriptions:users:title' => 'Notifications per user',
+	'notifications:subscriptions:users:description' => 'To receive notifications from your friends (on an individual basis) when they create new content, find them below and select the notification method you would like to use.',
 
+	'notifications:subscriptions:groups:title' => 'Group notifications',
```
should this not come from groups?

jdalsem commented on mod/notifications/views/default/forms/notifications/subscriptions/groups.php:
```diff
@@ -4,6 +4,10 @@
  *
  * @uses $vars['user'] ElggUser
  */
+
```
move to groups???

jeabakker replied:
yes not in the PR

jdalsem commented on mod/site_notifications/elgg-plugin.php:
```diff
@@ -72,4 +72,11 @@
+	'events' => [
+		'create' => [
+			'user' => [
+				'Elgg\SiteNotifications\Users::enableSiteNotifications' => [],
```
maybe register low priority?

jeabakker commented on engine/classes/Elgg/Comments/AutoSubscribeHandler.php:
```diff
@@ -0,0 +1,46 @@
+		
+		$owner = $object->getOwnerEntity();
+		$container = $object->getContainerEntity();
+		if (!$owner instanceof \ElggUser || !$container instanceof \ElggEntity) {
```
nope
